[build-script] Cross-compile sourcekit-lsp for arm64

ahoppen opened the pull request:
Currently, when building an open source toolchain, SourceKit-LSP is only built for x86_64. Copy the cross-compilation config from swiftpm.py to also produce a fat sourcekit-lsp executable for both x86_64 and arm64.

rdar://78039145

ahoppen commented:
https://github.com/apple/sourcekit-lsp/pull/556

@swift-ci Please build toolchain

buttaface commented on utils/swift_build_support/swift_build_support/products/indexstoredb.py:
```diff
@@ -74,6 +74,10 @@ def get_dependencies(cls):
 
+    @classmethod
+    def has_cross_compile_hosts(self, args):
+        return args.cross_compile_hosts
```
Maybe pull this into the base class so [all products can use it](https://github.com/apple/swift/blob/e9eb3b06c1e62d65062cdc44db7f33e673cf5345/utils/swift_build_support/swift_build_support/products/swiftdriver.py#L131)?

ahoppen replied:
Good idea :+1

buttaface commented on utils/swift_build_support/swift_build_support/products/indexstoredb.py:
```diff
@@ -105,18 +109,24 @@ def run_build_script_helper(action, host_target, product, args,
+                helper_cmd += [cross_compile_host]
+        elif product.is_cross_compile_target(host_target):
+            helper_cmd += ['--cross-compile-hosts', host_target,
+                           '--skip-cmake-bootstrap']
```
Please just copy the above lines here instead, as this is a SPM-specific flag.

finagolfin reviewed (changes requested):
Makes sense, but some changes needed.

ahoppen commented:
@swift-ci Please build toolchain macOS

ahoppen commented:
@swift-ci Please build toolchain macOS

ahoppen commented:
@swift-ci Please build toolchain macOS

ahoppen commented:
@swift-ci Please build toolchain macOS

ahoppen commented:
https://github.com/apple/sourcekit-lsp/pull/556

@swift-ci Please build toolchain macOS

ahoppen committed:
[build-script] Cross-compile sourcekit-lsp for arm64

Currently, when building an open source toolchain, SourceKit-LSP is only built for x86_64. Copy the cross-compilation config from swiftpm.py to also produce a fat sourcekit-lsp executable for both x86_64 and arm64.

rdar://78039145

ahoppen commented:
https://github.com/apple/sourcekit-lsp/pull/556

@swift-ci Please build toolchain macOS

ahoppen commented:
https://github.com/apple/sourcekit-lsp/pull/556

@swift-ci Please build toolchain macOS

ahoppen commented:
https://github.com/apple/sourcekit-lsp/pull/556

@swift-ci Please build toolchain macOS

ahoppen commented:
https://github.com/apple/sourcekit-lsp/pull/556

@swift-ci Please smoke test

ahoppen commented:
https://github.com/apple/sourcekit-lsp/pull/556

@swift-ci Please build toolchain Linux

ahoppen commented:
https://github.com/apple/sourcekit-lsp/pull/556

@swift-ci Please build toolchain Windows

benlangmuir commented on utils/swift_build_support/swift_build_support/products/indexstoredb.py:
```diff
@@ -60,6 +60,13 @@ def should_install(self, host_target):
+    def supports_cross_compilation(self):
+        """
+        Whether the product can be built for a different architecture than the 
+        host by passing --cross-compile-host to it.
```
Why not just override has_cross_compile_hosts on IndexStoreDB?

ahoppen replied:
Because I didn’t think of that. Good idea 👍

benlangmuir reviewed (approved):
Made a minor suggestion, but LGTM.

ahoppen commented:
https://github.com/apple/sourcekit-lsp/pull/556

@swift-ci Please smoke test
